core/src/main/java/hudson/model/Queue.java

```diff
@@ -1717,18 +1727,22 @@ private Runnable makeFlyWeightTaskBuildable(final BuildableItem p){
                     continue;
                 }
 
-                LOGGER.log(Level.FINEST, "Creating flyweight task {0} for computer {1}", new Object[]{fullDisplayName, c.getName()});
-                return new Runnable() {
-                    @Override public void run() {
-                        c.startFlyWeightTask(new WorkUnitContext(p).createWorkUnit(p.task));
-                        makePending(p);
-                    }
-                };
+                return createFlyWeightTaskRunnable(p, fullDisplayName, c);
             }
         }
         return null;
     }
 
+    private Runnable createFlyWeightTaskRunnable(final BuildableItem p, String fullDisplayName, final Computer c) {
+        LOGGER.log(Level.FINEST, "Creating flyweight task {0} for computer {1}", new Object[]{fullDisplayName, c.getName()});
```

Comment:
Suggest we adopt this approach on the PR for the displayName, and don't pass as a param: https://github.com/jenkinsci/jenkins/pull/2947/files#diff-7e8de48a31688987a2521079ff3828b9R1620